src/plugins/fs/vfat.c

```diff
@@ -328,6 +325,89 @@ gboolean bd_fs_vfat_check_label (const gchar *label, GError **error) {
     return TRUE;
 }
 
+/**
+ * bd_fs_vfat_set_uuid:
+ * @device: the device containing the file system to set uuid for
+ * @uuid: (nullable): volume ID to set or %NULL to generate a new one
+ * @error: (out) (optional): place to store error (if any)
+ *
+ * Returns: whether the volume ID of vfat file system on the @device was
+ *          successfully set or not
+ *
+ * Tech category: %BD_FS_TECH_VFAT-%BD_FS_TECH_MODE_SET_UUID
+ */
+gboolean bd_fs_vfat_set_uuid (const gchar *device, const gchar *uuid, GError **error) {
+    const gchar *args[5] = {"fatlabel", "-i", device, NULL, NULL};
+    g_autofree gchar *new_uuid = NULL;
+    size_t len = 0;
+
+    if (!check_deps (&avail_deps, DEPS_FATLABELUUID_MASK, deps, DEPS_LAST, &deps_check_lock, error))
+        return FALSE;
+
+    if (!uuid || g_strcmp0 (uuid, "") == 0)
+        args[3] = "--reset";
+    else {
+        len = strlen (uuid);
+        /* we want to support vol ID in the "udev format", e.g. "2E24-EC82" */
+        if (len == 9 && uuid[4] == '-') {
+            new_uuid = g_new0 (gchar, 9);
+            memcpy (new_uuid, uuid, 4);
+            memcpy (new_uuid + 4, uuid + 5, 4);
+        } else
+            new_uuid = g_strdup (uuid);
+
+        args[3] = new_uuid;
+    }
+
+    return bd_utils_exec_and_report_error (args, NULL, error);
+}
+
+/**
+ * bd_fs_vfat_check_uuid:
+ * @uuid: UUID to check
+ * @error: (out) (optional): place to store error
+ *
+ * Returns: whether @uuid is a valid UUID for the vfat file system or not
+ *          (reason is provided in @error)
+ *
+ * Tech category: always available
+ */
+gboolean bd_fs_vfat_check_uuid (const gchar *uuid, GError **error) {
+    guint64 vol_id;
+    gchar *new_uuid = NULL;
+    gchar *endptr = NULL;
+    size_t len = 0;
+
+    if (!uuid)
+        return TRUE;
+
+    len = strlen (uuid);
+    if (len == 9 && uuid[4] == '-') {
+        new_uuid = g_new0 (gchar, 9);
+        memcpy (new_uuid, uuid, 4);
+        memcpy (new_uuid + 4, uuid + 5, 4);
+    } else
+        new_uuid = g_strdup (uuid);
+
+    vol_id = g_ascii_strtoull (new_uuid, &endptr, 16);
+    if (vol_id == 0 || *endptr) {
```

Comment:
Hmm, is `0000-0000` valid UUID? I'd also suggest to check for non-NULL `endptr` before dereferencing it.

Reply:
> Hmm, is `0000-0000` a valid UUID?

Yes. Or at least `fatlabel` is not complaining about it and will set it.